UI/app/src/main/java/com/example/pathfinder/Activities/PassengerDashboard.java

```diff
@@ -127,9 +167,153 @@ public void onClick(View v) {
+                @Override
+                public void messageArrived(String topic, MqttMessage message) {
+                    if (topic.equals(NEXT_STOP)) {
+                        updateNextStop(message);
+                        //Here a comparison could be made to check is the next stop is the chosen one. If so, we an notify the user.
+                    } else if(topic.equals(BUS_STOP_LIST_TOPIC)){
+                        stopList = parseMessageToArray(message);
+                        generateBusLine();
+                    }else if(topic.equals(BUS_NAME_TOPIC)){
+                        generateBusName(message);
+                    }
+
+                }
+
+                @Override
+                public void deliveryComplete(IMqttDeliveryToken token) {
+                    Log.d(TAG, "Message delivered");
+                }
+            });
+        }
+    }
+
+
+
+
+    /**
+     * Method that checks if the mqtt broker is connected to the app
+     * create a toast if not notify the problem
+     * should be used before sending messages
+     */
+    private void notConnected() {
+        if (!isMqttConnected) {
+            final String notConnected = "No connection";
+            Log.e(TAG, notConnected);
+            Toast.makeText(getApplicationContext(), notConnected, Toast.LENGTH_SHORT).show();
+            return;
+        }
+    }
+
+    /**
+     * Update the interface with the next stop when receiving a new next stop from the driver's side.
+     * @param message -> mqtt message send on the relevant topic NEXT_STOP.
+     */
+    private void updateNextStop(MqttMessage message){
+        String nextStop = "Next stop: " + message.toString();
+        mNextStop.setText(nextStop);
+    }
+
 
+    /**
+     * Notify the driver that the passenger is connected to the bus line. Triggers the sending of a message containing all buss stops.
+     * Message need to be parsed.
+     * Please refer to parseMessageToArray() method.
+     */
+    private void notifyDriverAboutNewPassenger(){
+        mMqttClient.publish(NEW_PASSENGER, NEW_PASSENGER_BUS_ROUTE_TRIGGER, QOS, null);
+        Log.d("Passenger to Driver", "Request bus route" );
+        mMqttClient.publish(NEW_PASSENGER, NEW_PASSENGER_BUS_NAME_TRIGGER, QOS, null);
+        Log.d("Passenger to Driver", "Request bus name" );
     }
 
+    /**
+     * Parse the MQTT message into an array of strings. Characters need to be separated by a ";" in order to be separated during parsing.
+     * @param message -> The MQTT message to be parsed
+     * @return an array of strings.
+     */
+    private ArrayList<String> parseMessageToArray(MqttMessage message) {
+        ArrayList messageAsArray = new ArrayList();
+        String message_ = message.toString();
+        Log.d("Bus Stop", message_);
+        char[] charList = message_.toCharArray();
+        String stop = "";
+        for (char c : charList){
```

Comment:
Here you could use the method String.split() on something like
String[]  messageArray = message_.split(";");
To parse the message to a string array? So you won't need the for loop, I think :)

Reply:
Ok, I know why haha, I am using a ArrayList for the StopList adapter. and split returns an array, not an array list. So I have to convert the array into an array list afterwards or creating a new/adapt the AdapterClass to use the array, which was more work than parsing "manually"